chronos/apps/http/httpcommon.nim

```diff
@@ -0,0 +1,168 @@
+#
+#            Chronos HTTP/S common types
+#             (c) Copyright 2021-Present
+#         Status Research & Development GmbH
+#
+#              Licensed under either of
+#  Apache License, version 2.0, (LICENSE-APACHEv2)
+#              MIT license (LICENSE-MIT)
+import stew/results, httputils, strutils, uri
+import ../../asyncloop, ../../asyncsync
+import ../../streams/[asyncstream, boundstream]
+export results, httputils, strutils
+
+const
+  HeadersMark* = @[byte(0x0D), byte(0x0A), byte(0x0D), byte(0x0A)]
+  PostMethods* = {MethodPost, MethodPatch, MethodPut, MethodDelete}
+
+type
+  HttpResult*[T] = Result[T, string]
+  HttpResultCode*[T] = Result[T, HttpCode]
+
+  HttpDefect* = object of Defect
+  HttpError* = object of CatchableError
+  HttpCriticalError* = object of HttpError
+    code*: HttpCode
+  HttpRecoverableError* = object of HttpError
+    code*: HttpCode
+  HttpDisconnectError* = object of HttpError
+
+  TransferEncodingFlags* {.pure.} = enum
+    Identity, Chunked, Compress, Deflate, Gzip
+
+  ContentEncodingFlags* {.pure.} = enum
+    Identity, Br, Compress, Deflate, Gzip
+
+  HttpBodyReader* = ref object of AsyncStreamReader
+    streams*: seq[AsyncStreamReader]
+
+proc newHttpBodyReader*(streams: varargs[AsyncStreamReader]): HttpBodyReader =
+  ## HttpBodyReader is AsyncStreamReader which holds references to all the
+  ## ``streams``. Also on close it will close all the ``streams``.
+  ##
+  ## First stream in sequence will be used as a source.
+  doAssert(len(streams) > 0, "At least one stream must be added")
+  var res = HttpBodyReader(streams: @streams)
+  res.init(streams[0])
+  res
+
+proc closeWait*(bstream: HttpBodyReader) {.async.} =
+  ## Close and free resource allocated by body reader.
+  if len(bstream.streams) > 0:
+    var res = newSeq[Future[void]]()
+    for item in bstream.streams.items():
+      res.add(item.closeWait())
+    await allFutures(res)
+  await procCall(AsyncStreamReader(bstream).closeWait())
+
+proc atBound*(bstream: HttpBodyReader): bool {.
+     raises: [Defect].} =
+  ## Returns ``true`` if lowest stream is at EOF.
+  let lreader = bstream.streams[^1]
+  doAssert(lreader of BoundedStreamReader)
+  let breader = cast[BoundedStreamReader](lreader)
+  breader.atEof() and (breader.bytesLeft() == 0)
+
+proc newHttpDefect*(msg: string): ref HttpDefect {.
+     raises: [HttpDefect].} =
+  newException(HttpDefect, msg)
+
+proc newHttpCriticalError*(msg: string,
+                           code = Http400): ref HttpCriticalError {.
+     raises: [HttpCriticalError].} =
+  var tre = newException(HttpCriticalError, msg)
+  tre.code = code
+  tre
+
+proc newHttpRecoverableError*(msg: string,
+                              code = Http400): ref HttpRecoverableError {.
+     raises: [HttpRecoverableError].} =
+  var tre = newException(HttpRecoverableError, msg)
+  tre.code = code
+  tre
+
+proc newHttpDisconnectError*(): ref HttpDisconnectError {.
+     raises: [HttpDisconnectError].} =
```

Comment:
I'm not sure what the rationale for having `newXxx` functions is to begin with - perhaps they predate object init syntax and constructing an exception field by field using `result` was the least bad way to do things,  but an alternative is to simply use `(ref HttpDisconnectError)(msg: "..", otherfield: yyy)` without a special `newXxx` function.

One reason might be to save on c codegen - allocating and raising an exception is quite voluminous in terms of generated code - if the objective is to avoid all this code at the raise site, a more efficient style is to use `raiseXxxError` functions that also do the raising - this has benefits because assuming exceptions are rare, more of the fast path can fit in caches etc - the std library for example uses this style to raise its defects.

Basically, every time one uses `raise (ref ValueError)(...)`, the following blob of c code is generated, and this is for a simple `ValueError` that doesn't have any extra fields:
```
	tyObject_ValueError__yoNlBGx0D2tRizIdhQuENw* T1_;
	NimStringDesc* T2_;
	T1_ = (tyObject_ValueError__yoNlBGx0D2tRizIdhQuENw*)0;
	T1_ = (tyObject_ValueError__yoNlBGx0D2tRizIdhQuENw*) newObj((&NTI__Ie1m0dv1ZHg72IgPRr1cDw_), sizeof(tyObject_ValueError__yoNlBGx0D2tRizIdhQuENw));
	(*T1_).Sup.Sup.Sup.m_type = (&NTI__yoNlBGx0D2tRizIdhQuENw_);
	(*T1_).Sup.Sup.name = "ValueError";
	T2_ = (NimStringDesc*)0;
	T2_ = (*T1_).Sup.Sup.message; (*T1_).Sup.Sup.message = copyStringRC1(msg);
	if (T2_) nimGCunrefNoCycle(T2_);
	raiseExceptionEx((Exception*)T1_, "ValueError", "raiseValueError", "/home/arnetheduck/status/nim-beacon-chain/beacon_chain/testit.n"
"im", 2);
```

One important note: such functions need to be annotated with `noreturn` or compiler analysis breaks: `raiseXxx() {.noreturn.}`